applications/TrilinosApplication/custom_strategies/builder_and_solvers/trilinos_block_builder_and_solver.h

```diff
@@ -4,954 +4,830 @@
+     * vector
+     * @param pScheme The integration scheme considered
+     * @param rModelPart The model part of the problem to solve
+     * @param rA The LHS matrix
+     * @param rb The RHS vector
+     */
+    void Build(typename TSchemeType::Pointer pScheme,
+               ModelPart& rModelPart,
+               TSystemMatrixType& rA,
+               TSystemVectorType& rb) override
     {
         KRATOS_TRY
-        if(!pScheme)
-            KRATOS_ERROR << "No scheme provided!";
-
-        //getting the elements from the model
-        ElementsArrayType& pElements = r_model_part.Elements();
 
-        //getting the array of the conditions
-        ConditionsArrayType& ConditionsArray = r_model_part.Conditions();
+        KRATOS_ERROR_IF(!pScheme) << "No scheme provided!" << std::endl;
 
-        //resetting to zero the vector of reactions
+        const int nelements = static_cast<int>(rModelPart.Elements().size());
+        const int nconditions = static_cast<int>(rModelPart.Conditions().size());
+        // Resetting to zero the vector of reactions
         TSparseSpace::SetToZero(*BaseType::mpReactionsVector);
 
-        //contributions to the system
-        LocalSystemMatrixType LHS_Contribution = LocalSystemMatrixType(0,0);
+        // Contributions to the system
+        LocalSystemMatrixType LHS_Contribution = LocalSystemMatrixType(0, 0);
         LocalSystemVectorType RHS_Contribution = LocalSystemVectorType(0);
 
-        //vector containing the localization in the system of the different
-        //terms
-        Element::EquationIdVectorType EquationId;
-
-        //			int rank = A.Comm().MyPID(); //getting the processor Id
+        // vector containing the localization in the system of the different terms
+        Element::EquationIdVectorType equation_ids_vector;
+        ProcessInfo& r_current_process_info = rModelPart.GetProcessInfo();
+        const ModelPart::ElementsContainerType::ptr_iterator el_begin = rModelPart.Elements().ptr_begin();
+        const ModelPart::ConditionsContainerType::ptr_iterator cond_begin = rModelPart.Conditions().ptr_begin();
 
-        ProcessInfo& CurrentProcessInfo = r_model_part.GetProcessInfo();
         // assemble all elements
-        for (typename ElementsArrayType::ptr_iterator it=pElements.ptr_begin(); it!=pElements.ptr_end(); ++it)
-        {
-            //calculate elemental contribution
-            pScheme->CalculateSystemContributions(*it,LHS_Contribution,RHS_Contribution,EquationId,CurrentProcessInfo);
-
-            //assemble the elemental contribution
-            TSparseSpace::AssembleLHS(A,LHS_Contribution,EquationId);
-            TSparseSpace::AssembleRHS(b,RHS_Contribution,EquationId);
-
-            // clean local elemental memory
-            pScheme->CleanMemory(*it);
+        #pragma omp parallel for firstprivate(equation_ids_vector, LHS_Contribution, RHS_Contribution, r_current_process_info)
```

Comment:
this is still parallel...so you need to ensure that epetra is compiled with support for atomics

Reply:
so as per above comment, I will remove this and we can add it later in a different PR !